### PR TITLE
CHE-9840 add factory events handling

### PR DESCRIPTION
**Pull request by olexii4**
### What does this PR do?
Add factory events handling. Implement OpenFile action.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/10162
https://github.com/eclipse/che/issues/9840
	
https://bluejeans.com/s/s@Q@K/

Signed-off-by: Oleksii Orel <oorel@redhat.com>

### CODE REVIEW

File: che-theia-factory/src/browser/types.ts [hunk 1]
```diff
@@ -13,4 +13,25 @@ import { IWorkspaceConfig } from "@eclipse-che/workspace-client";
...
+
+export interface IFactoryAction {
+    id: string,
+    properties?: {
+        name?: string,
+        file?: string,
+        greetingTitle?: string,
+        greetingContentUrl?: string
+    }
+}
```
Comment (AndrienkoAleksandr) [inline che-theia-factory/src/browser/types.ts]: Missed line.

File: che-theia-factory/src/browser/factory-theia-client.ts [hunk 2]
```diff
@@ -52,31 +106,29 @@ export class FactoryTheiaClient implements FrontendApplicationContribution {
...
 
             const source = project.source;
-            const projectPath = projectsRoot + project.path;
+            const projectPath = this.projectsRoot + project.path;
 
-            this.messageService.info(`Cloning ... ${source.location} to ${projectPath}...`)
+            this.messageService.info(`Cloning ... ${source.location} to ${projectPath}...`);
 
-            this.git.clone(
+            importProjectPromices.push(this.git.clone(
```
Comment (AndrienkoAleksandr) [inline che-theia-factory/src/browser/factory-theia-client.ts]: We could use "await" here. But up to you.

File: che-theia-factory/src/browser/factory-theia-client.ts [hunk 3]
```diff
@@ -85,22 +137,39 @@ export class FactoryTheiaClient implements FrontendApplicationContribution {
...
                     this.messageService.info(`Project ${projectPath} successfully cloned.`);
                     if (source.parameters['branch']) {
-                        const options: Git.Options.Checkout.CheckoutBranch = { branch: source.parameters['branch'] };
+                        const options: Git.Options.Checkout.CheckoutBranch = {branch: source.parameters['branch']};
                         this.git.checkout(repo, options);
                     }
+                    return Promise.resolve();
                 }
             ).catch((error) => {
                 console.error(`Couldn't clone ${source.location} to ${projectPath}... ${error}`);
```
Comment (AndrienkoAleksandr) [inline che-theia-factory/src/browser/factory-theia-client.ts]: Minor, but such case: clone could be done successfully, but than error can be thrown on checkout operation. Message service always displays  error message "Couldn't clone...", but in this case it should only notify about checkout fail.

File: che-theia-factory/src/browser/factory-theia-client.ts [hunk 4]
```diff
@@ -31,19 +40,64 @@ const queryString = require('query-string');
...
+            this.appLoadedEmitter.fire({actions: this.onAppLoadedActions});
+        });
+
+        window.onbeforeunload = () => {
+            this.appClosedEmitter.fire({actions: this.onAppClosedActions});
+        };
+
+        this.onAppLoaded((event: { actions: Array<IFactoryAction> }) => {
+            event.actions.forEach((onAppLoadedAction: IFactoryAction) => {
+                console.log('>>>>>>>>>>> onAppLoaded.action', onAppLoadedAction);
```
Comment (benoitf) [inline che-theia-factory/src/browser/factory-theia-client.ts]: log trace to be removed ?

File: che-theia-factory/src/browser/factory-theia-client.ts [hunk 5]
```diff
@@ -31,19 +40,64 @@ const queryString = require('query-string');
...
+                    default:
+                        const errorMessage = `Action id '${onProjectsLoadedAction.id}' is not supported yet!`;
+                        console.error(errorMessage);
+                        this.messageService.error(errorMessage);
+                }
+            });
+        });
+        this.onAppClosed(() => {
+            this.onAppClosedActions.forEach((onAppClosedAction: IFactoryAction) => {
+                console.log('>>>>>>>>>>> onAppClosed.action', onAppClosedAction);
```
Comment (benoitf) [inline che-theia-factory/src/browser/factory-theia-client.ts]: log trace to remove ?

File: che-theia-factory/src/browser/factory-theia-client.ts [hunk 6]
```diff
@@ -85,22 +137,39 @@ export class FactoryTheiaClient implements FrontendApplicationContribution {
...
+        const uri = new URI().withPath(this.projectsRoot + relativePath).withScheme('file');
+        await this.editorManager.open(uri);
     }
 
     getEnvVariable(name: string): EnvVariable | undefined {
         if (!this.envVariables) {
             return undefined;
         }
-        return this.envVariables.find(function(envVariable) { return envVariable.name === name });
+        return this.envVariables.find(function (envVariable) {
```
Comment (benoitf) [inline che-theia-factory/src/browser/factory-theia-client.ts]: use of arrow function instead ?

**Review by sunix — CHANGES_REQUESTED** [review]
missing 
in che-theia-factory/package.json ? 
```
"@theia/editor": "0.3.12",
```

**Comment by sunix** [comment]
could you also rebase ? (so it works on Theia 0.3.12)

File: che-theia-factory/src/browser/types.ts (hunk 1 above)
Comment (olexii4) [reply che-theia-factory/src/browser/types.ts]: done

File: che-theia-factory/src/browser/factory-theia-client.ts (hunk 2 above)
Comment (olexii4) [reply che-theia-factory/src/browser/factory-theia-client.ts]: I am sure about that. We should start in parallel and wait for resolve importProjectPromices.

File: che-theia-factory/src/browser/factory-theia-client.ts (hunk 3 above)
Comment (olexii4) [reply che-theia-factory/src/browser/factory-theia-client.ts]: Good catch. Done

File: che-theia-factory/src/browser/factory-theia-client.ts (hunk 4 above)
Comment (olexii4) [reply che-theia-factory/src/browser/factory-theia-client.ts]: Done

File: che-theia-factory/src/browser/factory-theia-client.ts (hunk 5 above)
Comment (olexii4) [reply che-theia-factory/src/browser/factory-theia-client.ts]: Done

File: che-theia-factory/src/browser/factory-theia-client.ts (hunk 6 above)
Comment (olexii4) [reply che-theia-factory/src/browser/factory-theia-client.ts]: Done

**Comment by olexii4** [comment]
Done

**Review by AndrienkoAleksandr — APPROVED** [review]

File: che-theia-factory/src/browser/factory-manager.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
+/*
```
Comment (benoitf) [inline che-theia-factory/src/browser/factory-manager.ts]: FYI license there https://github.com/eclipse/che-theia-factory-extension/blob/master/LICENSE is EPL 2.0

so copyright headers should be 2.0 as well.

**Review by sunix — APPROVED** [review]

File: che-theia-factory/src/browser/che-theia-factory-frontend-module.ts [hunk 8]
```diff
@@ -1,7 +1,7 @@
 /*
  * Copyright (c) 2018-2018 Red Hat, Inc.
  * All rights reserved. This program and the accompanying materials
- * are made available under the terms of the Eclipse Public License v1.0
+ * are made available under the terms of the Eclipse Public License v2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-v10.html
```
Comment (benoitf) [inline che-theia-factory/src/browser/che-theia-factory-frontend-module.ts]: link is still pointing to v10.html

example of EPL 2.0 headers:
https://github.com/eclipse/theia-generator-plugin/blob/master/src/app/index.ts#L1-L9

Comment (benoitf) [reply che-theia-factory/src/browser/che-theia-factory-frontend-module.ts]: FAQ being there
https://www.eclipse.org/legal/epl-2.0/faq.php#h.q72cnghf29k0

**Comment by benoitf** [comment]
headers should be 
```
/*********************************************************************
* Copyright (c) 2018 Red Hat, Inc.
*
* This program and the accompanying materials are made
* available under the terms of the Eclipse Public License 2.0
* which is available at https://www.eclipse.org/legal/epl-2.0/
*
* SPDX-License-Identifier: EPL-2.0
**********************************************************************/
```
https://www.eclipse.org/legal/epl-2.0/faq.php#h.q72cnghf29k0

**Comment by olexii4** [comment]
Done

**Review by akurinnoy — APPROVED** [review]

**Review by benoitf — APPROVED** [review]
thx

**Pull request merged**